Proposal.md

```diff
@@ -0,0 +1,289 @@
+
+### HPA biased
+In case both HPA and VPA have recommendations to scale, HVPA controller will always apply HPA's recommendations. VPA's recommendation will be applied only when HPA doesn't have any recommendation.
+
+#### Pros:
+* Works even if HPA and VPA act on different metrices
+* Vertical scaling is done only when required
+#### Cons:
+* Might need both VPA and HPA to be deployed in a recommendation-only mode.
+
+### VPA biased
+In case both HPA and VPA have recommendations to scale, HVPA controller will always apply VPA's recommendations. HPA's recommendation will be applied only when VPA doesn't have any recommendation.
+
+#### Pros:
+* Works even if HPA and VPA act on different metrices
+#### Cons:
+* May have more frequent rolling updates. Can it be mitigated by setting user provided thresholds for vertical autoscaling?
+* Does it make sense to give preference to VPA instead of HPA, when vertical scaling may result in rolling updates?
+* Will need both VPA and HPA to be deployed in a recommendation-only mode.
+
+### Mixed
+In this mode, along with the preference for autoscaling (HPA or VPA), user will also provide thresholds for horizontal and vertical autoscaling, say, `x` and `y` respectively. If only one autoscaler has a new recommendation, it will be applied. If both the components have recommendations:
+* HPA's recommendation will be applied if the difference between the recommended replicas and the current replicas is greater than `x`.
+* VPA's recommendation will be applied if the difference between the recommended value and the current value is greater than `y`, or the recommended value is more than the `limit`.
+* In case both above conditions are satisifed, user provided preference will be taken into account
+
+#### Pros:
+* Works even if HPA and VPA act on different metrices
+* Disruptions because of rolling updates can be controlled as rolling updates will be applied less frequently because of higher threshold
+* Provides more control to users for autoscaling apps
+* Relies on users' understanding of application, and its scaling
+#### Cons:
+* Relies on users' understanding of application, and its scaling
+* Will need both VPA and HPA to be deployed in a recommendation-only mode.
+
+### Weight based scaling
+In this mode, deployment's scaling is divided into 3 stages depending on number of replicas, and a user can provide an interval `x1` to `x2` and 3 `vpaWeight`s between and including `0` and `1`. Stage 1 will be from HPA's `minReplicas` to `x1`. Stage 2 will be from `x1` to `x2`. Stage 3 will be from `x2` to HPA's `maxReplicas`. HVPA controller will consider VPA's recommendations according to the weights provided, and scale the deployment accordingly. The weights given to the HPA's recommendation will be `1 - vpaWeight`.
+
+According to VPA, new `requests` will be `currentRequest + (targetRequest - currentRequest) * vpaWeight`
```

Comment:
Should we have one more parameter alpha[0,1]
According to VPA, new requests will be currentRequest + (targetRequest - currentRequest) * vpaWeight * alpha

According to HPA, new number of replicas will be ceil of currentReplicas + (desiredReplicas - currentReplicas) * (1 - hpaWeight) * (1 - alpha)

Then we have [hpa/vpa]Weight for how much confidence we should provide for the corresponding reccomendations from hpa/vpa and use alpha for toggling between hpa and vpa decisions.